File: algopt/rebalancer/examples/sudoku/Sudoku.cpp
// Copyright (c) Meta Platforms, Inc. and affiliates.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "algopt/rebalancer/interface/ProblemSolverFactory.h"

#include "fmt/core.h"
#include <folly/container/irange.h>
#include <folly/init/Init.h>
#include <folly/logging/xlog.h>
#include <folly/String.h>

#include <bitset>
#include <map>
#include <sstream>
#include <string>
#include <vector>

using namespace facebook::rebalancer::interface;

constexpr folly::StringPiece kDimensionsApproach{"dimensions"};
constexpr folly::StringPiece kPartitionsApproach{"partitions"};

DEFINE_string(
    approach,
    kPartitionsApproach.str(),
    ("What approach to use for solving the problem: "
     "partitions and group count constraints on partition groups "
     "or capacity constraints on dimensions"));

static void add_initial_assignment(
    ProblemSolver& solver,
    const std::vector<std::string>& puzzle) {
  // compute initial state to set for rebalancer
  std::map<std::string, std::vector<std::string>> initial_board;
  // also compute which numbers should not be moving
  std::vector<std::string> avoid_moving;

  // keep track of pre-assigned numbers
  std::bitset<10> numbers;

  // used to store all non-assigned numbers
  initial_board["dummy_square"] = {};
  for (const auto row : folly::irange(9)) {
    numbers.reset();
    for (const auto col : folly::irange(9)) {
      const auto& square = fmt::format("square_{}_{}", row, col);
      initial_board[square] = {};
      auto value = puzzle[row].at(col);
      if (value == '.') {
        continue;
      }
      int given_number = value - '0';
      XLOG_IF(FATAL, given_number < 1 || given_number > 9)
          << fmt::format("given value: {} is not a valid digit", value);
      const auto& number_label =
          fmt::format("number_{}_in_row_{}", given_number, row);
      initial_board[square].emplace_back(number_label);
      avoid_moving.emplace_back(number_label);
      numbers.set(given_number);
    }
    // all numbers which are not assigned to squares are assigned to a special
    // dummy square
    for (const auto number : folly::irange(1, 10)) {
      if (numbers.test(number)) { // already assigned, so continue
        continue;
      }
      initial_board["dummy_square"].emplace_back(
          fmt::format("number_{}_in_row_{}", number, row));
    }
  }

  // set initial assignment
  solver.setAssignment(initial_board);
  // avoid moving assigned squares
  AvoidMovingSpec avoidMovingSpecs;
  avoidMovingSpecs.objects() = avoid_moving;
  solver.addConstraint(avoidMovingSpecs);

  // completely free the dummy square
  ToFreeSpec toFreeSpec;
  toFreeSpec.containers() = {"dummy_square"};
  solver.addConstraint(toFreeSpec);
}

static void add_scopes(ProblemSolver& solver) {
  // define scopes: rows, columns and 3x3 boxes
  std::map<std::string, std::string> rows, columns, boxes;
  for (const auto row : folly::irange(9)) {
    for (const auto col : folly::irange(9)) {
      auto square_label = fmt::format("square_{}_{}", row, col);
      rows[square_label] = fmt::format("row_{}", row);
      columns[square_label] = fmt::format("col_{}", col);
      boxes[square_label] = fmt::format("box_{}_{}", row / 3, col / 3);
    }
  }
  solver.addScope("rows", rows);
  solver.addScope("columns", columns);
  solver.addScope("boxes", boxes);
}

static void use_partition_groups(ProblemSolver& solver) {
  // add partition group: all 'numbers' with same mathematical (numerical)
  // value belong to one partition group (part of the 'same_numbers' partition)
  std::map<std::string, std::string> same_numbers;
  for (const auto number : folly::irange(1, 10)) {
    for (const auto row : folly::irange(9)) {
      same_numbers[fmt::format("number_{}_in_row_{}", number, row)] =
          fmt::format("number_{}", number);
    }
  }
  solver.addPartition("same_numbers", std::move(same_numbers));

  // with in any scope item each partition group can have at most a count of 1
  constexpr std::array<std::string_view, 3> scopes = {
      "rows", "columns", "boxes"};
  for (const auto& scope : scopes) {
    GroupCountSpec groupCountSpec;
    groupCountSpec.scope() = scope;
    groupCountSpec.partitionName() = "same_numbers";

    Limit limit = Limit();
    limit.type() = LimitType::ABSOLUTE;
    limit.globalLimit() = 1;
    groupCountSpec.limit() = limit;

    solver.addConstraint(groupCountSpec);
  }
}

static void use_dimensions(ProblemSolver& solver) {
  // declare dimensions "dimension_{n}", iff number n it has value 1.
  for (const auto number : folly::irange(1, 10)) {
    std::map<std::string, double> dimensions;
    for (const auto row : folly::irange(9)) {
      dimensions[fmt::format("number_{}_in_row_{}", number, row)] = 1;
    }
    solver.addObjectDimension(fmt::format("dimension_{}", number), dimensions);
  }

  // only one object can have a non-zero dimension value in each scope
  constexpr std::array<std::string_view, 3> scopes = {
      "rows", "columns", "boxes"};
  for (const auto& scope : scopes) {
    for (const auto number : folly::irange(1, 10)) {
      CapacitySpec capacitySpec;
      capacitySpec.scope() = scope;
      capacitySpec.dimension() = fmt::format("dimension_{}", number);

      auto limit = Limit();
      limit.globalLimit() = 1;
      capacitySpec.limit() = limit;

      solver.addConstraint(capacitySpec);
    }
  }
}

static void print_solution(const AssignmentSolution& solution) {
  std::vector<std::vector<int>> solved_puzzle(9, std::vector<int>(9));
  std::vector<std::string_view> square_parts(5), number_parts(5);
  for (const auto& [number_label, square] : *solution.assignment()) {
    number_parts.clear();
    folly::split('_', number_label, number_parts);
    square_parts.clear();
    folly::split('_', square, square_parts);
    solved_puzzle[folly::to<int>(square_parts[1])]
                 [folly::to<int>(square_parts[2])] =
                     folly::to<int>(number_parts[1]);
  }

  std::stringstream ss;
  for (const auto row : folly::irange(9)) {
    for (const auto col : folly::irange(9)) {
      ss << solved_puzzle[row][col] << " ";
    }
    ss << std::endl;
  }
  XLOG(INFO) << "Final solution: \n" << ss.str();
}

static void solve_sudoku(
    const std::vector<std::string>& puzzle,
    const std::string& approach) {
  auto solver =
      ProblemSolverFactory::makeProblemSolver("rebalancer", "examples");
  // This is used to define the default dimension (has value 1 for all
  // objects)
  solver->setObjectName("number");
  // This is used to define a default scope (that has all containers)
  solver->setContainerName("square");

  add_initial_assignment(*solver, puzzle);
  add_scopes(*solver);

  if (approach == kPartitionsApproach) {
    use_partition_groups(*solver);
  } else if (approach == kDimensionsApproach) {
    use_dimensions(*solver);
  } else {
    XLOG(FATAL) << fmt::format(
        "{} must be either {} or {}",
        approach,
        kPartitionsApproach,
        kDimensionsApproach);
  }

  // each square should limit to have exactly one number
  CapacitySpec capacitySpec;
  capacitySpec.scope() = "square";
  capacitySpec.dimension() = "number_count";
  capacitySpec.limit()->globalLimit() = 1;

  solver->addConstraint(capacitySpec);

  OptimalSolverSpec solver_spec;
  solver_spec.solverPackage() = OptimalSolverPackage::HIGHS;
  solver->addSolver(solver_spec);

  auto solution = solver->solve();

  if (*solution.finalObjective()->value() > 0.01) {
    XLOG(FATAL) << "failed to solve sudoku completely";
  }

  print_solution(solution);
}

int main(int argc, char** argv) {
  const folly::Init init(&argc, &argv);

  const std::vector<std::string> puzzle = {
      "85...24..",
      "72......9",
      "..4......",
      "...1.7..2",
      "3.5...9..",
      ".4.......",
      "....8..7.",
      ".17......",
      "....36.4.",
  };
  solve_sudoku(puzzle, FLAGS_approach);

  return 0;
}

File: algopt/rebalancer/examples/sudoku/sudoku.py
#!/usr/bin/env python3
# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


# pyre-strict

from algopt.rebalancer.interface.py_client.ProblemSolver import ProblemSolver
from rebalancer.interface.thrift.Types.thrift_types import AssignmentSolution
from rebalancer.interface.thrift.v2.ProblemSolver.thrift_types import (
    ManifoldBackupParams,
    ManifoldUploadPolicy,
)
from rebalancer.interface.thrift.v2.ProblemSpecs.thrift_types import (
    AvoidMovingSpec,
    CapacitySpec,
    Limit,
)
from rebalancer.interface.thrift.v2.SolverSpecs.thrift_types import (
    OptimalSolverPackage,
    OptimalSolverSpec,
)


# See tutorial at:
# https://www.internalfb.com/intern/wiki/ReBalancer/API/Tutorial/


def sudoku_initial_assignment(solver: ProblemSolver) -> None:
    puzzle = [
        "85...24..",
        "72......9",
        "..4......",
        "...1.7..2",
        "3.5...9..",
        ".4.......",
        "....8..7.",
        ".17......",
        "....36.4.",
    ]

    # Compute initial board assignment.
    initial_board = {}
    # Also compute which numbers should not be moved.
    avoid_moving = []
    for row in range(0, 9):
        numbers = set(range(1, 10))  # keep track of unused numbers in this row
        for col in range(0, 9):
            initial_board["square_{}_{}".format(row, col)] = []  # empty
            if puzzle[row][col] != ".":
                given_number = int(puzzle[row][col])
                number_label = "number_{}_in_row_{}".format(given_number, row)
                initial_board["square_{}_{}".format(row, col)].append(number_label)
                avoid_moving.append(number_label)
                numbers.remove(given_number)
        # for the remaining numbers we can just put all of them on a single
        # square in the first column (they will be redistributed by the solver)
        initial_board["square_{}_0".format(row)].extend(
            "number_{}_in_row_{}".format(number, row) for number in numbers
        )

    solver.setAssignment(initial_board)

    spec = AvoidMovingSpec(name="avoid_moving", objects=avoid_moving)
    solver.addConstraint(spec)


def sudoku_add_dimensions(solver: ProblemSolver) -> None:
    # Declare dimensions "dimension_{n}", for a number n it has value 1.
    for number in range(1, 10):
        dimensions = {}
        for row in range(9):
            # addObjectDimension expects an iterable for the value in this map
            dimensions["number_{}_in_row_{}".format(number, row)] = 1
        # pyrefly: ignore [bad-argument-type]
        solver.addObjectDimension("dimension_{}".format(number), dimensions)


def sudoku_add_scopes(solver: ProblemSolver) -> None:
    # Define scopes
    rows = {}  # each row
    columns = {}  # each column
    boxes = {}  # each 3x3 boxes (there are 9 in total)
    for row in range(9):
        for col in range(9):
            square_label = "square_{}_{}".format(row, col)
            rows[square_label] = "row_{}".format(row)
            columns[square_label] = "col_{}".format(col)
            boxes[square_label] = "box_{}_{}".format(row // 3, col // 3)
    # Add scopes to the solver
    solver.addScope("rows", rows)
    solver.addScope("columns", columns)
    solver.addScope("boxes", boxes)


def sudoku_add_constraints(solver: ProblemSolver) -> None:
    limit = Limit(globalLimit=1)

    # Define constraints that limit each dimension to 1.
    # (only one object that has a non-zero dimension value can be
    # present in each scope)
    for scope_name in ["rows", "columns", "boxes"]:
        for number in range(1, 10):
            capacity_spec = CapacitySpec(
                name=f"capacity_{scope_name}_{number}",
                scope=scope_name,
                limit=limit,
                dimension="dimension_{}".format(number),
            )
            solver.addConstraint(capacity_spec)

    # Last but not least, each square should limit to
    # only one number in the square.
    capacity_spec = CapacitySpec(
        name="capacity",
        limit=limit,
        # Default scope defined for all containers (we didn't define it)
        scope="square",
        # Default dimension defined for all objects (we didn't define it)
        dimension="number_count",
    )
    solver.addConstraint(capacity_spec)


def main() -> None:
    solver = ProblemSolver(service_name="rebalancer", service_scope="examples")
    # This is used to define the default dimension (has value 1 for all objects)
    solver.setObjectName("number")
    # This is used to define a default scope (that has all containers)
    solver.setContainerName("square")

    sudoku_initial_assignment(solver)
    sudoku_add_dimensions(solver)
    sudoku_add_scopes(solver)
    sudoku_add_constraints(solver)

    solver_spec = OptimalSolverSpec(solverPackage=OptimalSolverPackage.HIGHS)
    solver.addSolver(solver_spec)

    backupParams = ManifoldBackupParams(uploadPolicy=ManifoldUploadPolicy.ON_FAILURE)
    solver.setManifoldBackupParams(backupParams)

    solution: AssignmentSolution = solver.solve()

    # Display a list of individual moves solver made
    # pyre-fixme[16]: Optional type has no attribute `__iter__`.
    moves = [move for summary in solution.movesSummary for move in summary.moves]
    assert len(moves) > 5
    for move in moves:
        print(
            "Moving {} from {} to {}".format(
                move.object.split("_")[1], move.srcContainer, move.dstContainer
            )
        )

    # Computation is complete, produce pretty output
    output = [[0 for col in range(9)] for row in range(9)]
    for number_label, square_label in solution.assignment.items():
        number = int(number_label.split("_")[1])
        row, col = square_label.split("_")[1:3]
        output[int(row)][int(col)] = number

    print("Final board:")
    for row in range(0, 9):
        print(output[row])

    if solution.finalObjective.value > 0.01:
        raise Exception("failed to solve completely")


if __name__ == "__main__":
    main()
